### PR TITLE
Update clearpiperun to use raw SQL

### LINKED ISSUE #690: clearpiperun consuming all available memory
Turns out that in order to delete the run, django is likely loading **all objects** into memory.

Currently the deletion logic is a simple:

https://github.com/askap-vast/vast-pipeline/blob/f2c207cd1b40e9a5e50ecba6e9c6468fa3da393d/vast_pipeline/management/commands/clearpiperun.py#L115

and with the cascading effect this is likely causing all sources and image objects to be loaded into memory.

Fix will be to break down the logic to delete components over a generated list.

This is likely the cause of the woes of Nimbus over the past few months because of the stuck command in the qcluster queue.

### PR DESCRIPTION
**Pull request by ddobie**
Fix #690.

### CODE REVIEW

**Comment by ddobie** [comment]
This now passes test, but I'm not convinced that my method of checking whether an image/skyregions are associated with more than one is actually efficient. It does work though.

File: vast_pipeline/management/commands/clearpiperun.py [hunk 1]
```diff
@@ -97,25 +95,30 @@ def handle(self, *args, **options) -> None:
...
+            logger.info('Clearing all pipeline run in the database')
             piperuns = list(Run.objects.values_list('name', flat=True))
 
         for piperun in piperuns:
             p_run_name = get_p_run_name(piperun)
             try:
                 p_run = Run.objects.get(name=p_run_name)
             except Run.DoesNotExist:
-                raise CommandError(f'Pipeline run {p_run_name} does not exist')
+                raise CommandError("Pipeline run '%s' does not exist", p_run_name)
```
Comment (mauch) [inline vast_pipeline/management/commands/clearpiperun.py:108]: You can leave this f-string - Its only the ones in `logging` calls that have issues :) 
```suggestion
                raise CommandError(f'Pipeline run {p_run_name} does not exist')
```

File: vast_pipeline/utils/delete_run.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+logger = logging.getLogger(__name__)
+
+def _run_raw_sql(command, cursor, debug=False, log=True):
+    if log:
+        log_str = f"Running {command}"
+        if debug:
+            logger.debug(log_str)
+        else:
+            logger.info(log_str)
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:14]:
```suggestion
        if debug:
            logger.debug("Running %s", command)
        else:
            logger.info("Running %s", command)
```

File: vast_pipeline/utils/delete_run.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        #_run_raw_sql(sql_cmd, cursor)
+        
+        # Fetch source IDs associated with the pipeline run
+        sql_cmd = f"SELECT id FROM vast_pipeline_source WHERE run_id = {p_run_id};"
+        _run_raw_sql(sql_cmd, cursor)
+        source_ids = cursor.fetchall()
+
+        # Iterate over each source ID and delete related information
+        n_source_ids = len(source_ids)
+        logger.info(f"Iterating over {n_source_ids} sources to delete tags and relations")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:34]:
```suggestion
        logger.info("Iterating over %d sources to delete tags and relations", n_source_ids)
```

File: vast_pipeline/utils/delete_run.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        _run_raw_sql(sql_cmd, cursor)
+
+        # Fetch image IDs associated with the pipeline run
+        sql_cmd = f"SELECT image_id FROM vast_pipeline_image_run WHERE run_id = {p_run_id};"
+        _run_raw_sql(sql_cmd, cursor)
+        image_ids = cursor.fetchall()
+
+        # Iterate over each image ID and delete related information
+        n_image_ids = len(image_ids)
+        logger.info(f"Iterating over {n_image_ids} images to delete measurements and images")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:89]:
```suggestion
        logger.info("Iterating over %d images to delete measurements and images", n_image_ids)
```

File: vast_pipeline/utils/delete_run.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+            _run_raw_sql(sql_cmd, cursor)
+            num_occurences = cursor.fetchone()[0]
+            
+            # Delete the link between the run and the image
+            sql_cmd = f"DELETE FROM vast_pipeline_image_run WHERE image_id = {image_id} AND run_id = {p_run_id};"
+            _run_raw_sql(sql_cmd, cursor, log=True)
+            
+            # If the image is associated with more than one run, do not delete the image.
+            if num_occurences > 1:
+                logger.debug("image_id %d is referenced by {num_occurences} other pipeline runs, not deleting", image_id)
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:106]:
```suggestion
                logger.debug("image_id %d is referenced by %d other pipeline runs, not deleting", image_id, num_occurences)
```

File: vast_pipeline/utils/delete_run.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                continue
+
+            try:
+                sql_cmd = f"DELETE FROM vast_pipeline_measurement WHERE image_id = {image_id};"
+                _run_raw_sql(sql_cmd, cursor, log=True)
+            except Exception as e:
+                print("++++++++++++++++++++++++++++++++")
+                print(e, image_id)
+                print("++++++++++++++++++++++++++++++++")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:115]: Rather use logger.error
```suggestion
                logger.error("%s %d", e, image_id)
```

File: vast_pipeline/utils/delete_run.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                pass
+
+            try:
+                sql_cmd = f"DELETE FROM vast_pipeline_image WHERE id = {image_id};"
+                _run_raw_sql(sql_cmd, cursor, log=True)
+            except Exception as e:
+                print("++++++++++++++++++++++++++++++++")
+                print(e, image_id)
+                print("++++++++++++++++++++++++++++++++")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:124]:
```suggestion
                logger.error("%s %d", e, image_id)
```

File: vast_pipeline/utils/delete_run.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+                pass
+        logger.info(f"Time to iterate over {n_image_ids} image ids: {timer.reset()} seconds")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:126]:
```suggestion
        t = timer.reset()
        logger.info("Time to iterate over %d image ids: %f seconds", n_image_ids, t)
```

File: vast_pipeline/utils/delete_run.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+        # Fetch skyregion IDs associated with the pipeline run
+        sql_cmd = f"SELECT skyregion_id FROM vast_pipeline_skyregion_run WHERE run_id = {p_run_id};"
+        _run_raw_sql(sql_cmd, cursor, debug=True)
+        sky_ids = cursor.fetchall()
+
+        # Iterate over each skyregion ID and delete related information
+        n_sky_ids = len(sky_ids)
+        logger.info(f"Iterating over {n_sky_ids} skyregion IDs to delete skyregions")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:135]:
```suggestion
        logger.info("Iterating over %d skyregion IDs to delete skyregions", n_sky_ids)
```

File: vast_pipeline/utils/delete_run.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+            sql_cmd = f"SELECT COUNT(*) FROM vast_pipeline_skyregion_run WHERE skyregion_id={sky_id};"
+            _run_raw_sql(sql_cmd, cursor)
+            num_occurences = cursor.fetchone()[0]
+            
+            sql_cmd = f"DELETE FROM vast_pipeline_skyregion_run WHERE skyregion_id = {sky_id} AND run_id = {p_run_id};"
+            _run_raw_sql(sql_cmd, cursor, log=True)
+            
+            # If the skyregion is associated with more than one run, do not delete the skyregion.
+            if num_occurences > 1:
+                logger.debug("skyregion_id %d is referenced by {num_occurences} other pipeline runs, not deleting", image_id)
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:150]:
```suggestion
                logger.debug("skyregion_id %d is referenced by %d other pipeline runs, not deleting", image_id, num_occurences)
```

File: vast_pipeline/utils/delete_run.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+                continue
+            
+            try:
+                sql_cmd = f"DELETE FROM vast_pipeline_skyregion WHERE id = {sky_id};"
+                _run_raw_sql(sql_cmd, cursor, log=True)
+            except Exception as e:
+                print("++++++++++++++++++++++++++++++++")
+                print(e, sky_id)
+                print("++++++++++++++++++++++++++++++++")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:159]:
```suggestion
                logger.error("%s %d", e, sky_id)
```

File: vast_pipeline/utils/delete_run.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+                pass
+        logger.info(f"Time to iterate over {n_sky_ids} sky ids: {timer.reset()} seconds")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:161]:
```suggestion
        t = timer.reset()
        logger.info("Time to iterate over %d sky ids: %f seconds", n_sky_ids, t)
```

File: vast_pipeline/utils/delete_run.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+            _run_raw_sql(sql_cmd, cursor, log=False)
+            sql_cmd = f"DELETE FROM vast_pipeline_relatedsource WHERE to_source_id = {source_id};"
+            _run_raw_sql(sql_cmd, cursor, log=False)
+
+            sql_cmd = f"DELETE FROM vast_pipeline_association WHERE source_id = {source_id};"
+            _run_raw_sql(sql_cmd, cursor, log=False)
+            
+            if i % 1000 == 0:
+                logger.info(f"Finished source id {source_id} ({i} of {n_source_ids})")
+        logger.info(f"Time to iterate over {n_source_ids} source ids: {timer.reset()} seconds")
```
Comment (mauch) [inline vast_pipeline/utils/delete_run.py:68]:
```suggestion
                logger.info("Finished source id %d (%d of %d)", source_id, i, n_source_ids)
        t = timer.reset()
        logger.info(f"Time to iterate over %d source ids: %f seconds", n_source_ids, t)
```

File: vast_pipeline/management/commands/clearpiperun.py [hunk 14]
```diff
@@ -97,25 +95,30 @@ def handle(self, *args, **options) -> None:
...
-            logger.info("Deleting pipeline '%s' from database", p_run_name)
+            logger.info("Deleting pipeline run '%s' from database", p_run_name)
             with transaction.atomic():
                 p_run.status = 'DEL'
                 p_run.save()
-            p_run.delete()
+
+            timer.reset()
+            delete_pipeline_run_raw_sql(p_run)
+            logger.info("Time to delete run from database: %.2f sec", timer.reset())
```
Comment (mauch) [inline vast_pipeline/management/commands/clearpiperun.py:117]:
```suggestion
            t = timer.reset()
            logger.info("Time to delete run from database: %.2f sec", t)
```

File: vast_pipeline/management/commands/clearpiperun.py [hunk 15]
```diff
@@ -97,25 +95,30 @@ def handle(self, *args, **options) -> None:
...
 
             # remove forced measurements in db if presents
             forced_parquets = remove_forced_meas(p_run.path)
+            logger.info("Time to delete forced measurements: %.2f sec", timer.reset())
```
Comment (mauch) [inline vast_pipeline/management/commands/clearpiperun.py:121]:
```suggestion
            t = timer.reset()
            logger.info("Time to delete forced measurements: %.2f sec", t)
```

File: vast_pipeline/management/commands/clearpiperun.py [hunk 16]
```diff
@@ -132,6 +135,7 @@ def handle(self, *args, **options) -> None:
                             f'Parquet file "{os.path.basename(parquet)}" not existent'
                         ))
                         pass
+            logger.info("Time to delete parquet files: %.2f sec", timer.reset())
```
Comment (mauch) [inline vast_pipeline/management/commands/clearpiperun.py:138]:
```suggestion
            t = timer.reset()
            logger.info("Time to delete parquet files: %.2f sec", t)
```

**Review by mauch — APPROVED** [review]
Mostly just lots of little f-string in log niggles :)
I tested and it works fine for me - you should test again to make sure my suggested changes are correct.

**Comment by ddobie** [comment]
Thanks! Will give it another test locally before merging just to make sure :)

**Pull request merged**